cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -130,13 +134,38 @@ private void publish(ProgramRunId programRunId, ImmutableMap.Builder<String, Str
     // ProgramRunId is always required in a notification
     properties.put(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId));
     Notification programStatusNotification = new Notification(Notification.Type.PROGRAM_STATUS, properties.build());
-    try {
-      messagingService.publish(StoreRequestBuilder.of(topicId)
-                                                  .addPayloads(GSON.toJson(programStatusNotification))
-                                                  .build()
-      );
-    } catch (Exception e) {
-      LOG.warn("Error while publishing notification for program {}: {}", programRunId, e);
+
+    int failureCount = 0;
+    long startTime = -1L;
+    boolean done = false;
+    // TODO CDAP-12255 this code was basically copied from MessagingMetricsCollectionService.TopicPayload#publish.
+    // This should be refactored into a common class for publishing to TMS with a retry strategy
+    while (!done) {
+      try {
+        messagingService.publish(StoreRequestBuilder.of(topicId)
+                                   .addPayloads(GSON.toJson(programStatusNotification))
+                                   .build());
+        done = true;
+      } catch (Exception e) {
```

Comment:
`messagingService.publish(...)`. In fact, the interface shows that it doesn't throw `ServiceUnavailableException` so there is no point in catching that?